File: shared/research.ts
// Data-driven Blacksmith research. Pure: numbers/defs in, numbers/defs out — no
// SpacetimeDB or Three deps — so the authoritative module, the client UI and the
// tests all fold the SAME deltas. Bonuses are NEVER baked onto unit rows; they
// are DERIVED on read from the owner's completed-tech bitmask, so a researched
// upgrade applies to every CURRENT and FUTURE unit automatically.

import { BuildingKind, ArmorClass } from './enums.ts';
import type { UnitDef } from './units.ts';
import { UNIT_DEFS } from './units.ts';
import type { BuildingDef } from './buildings_defs.ts';
import { BUILDING_DEFS } from './buildings_defs.ts';
import type { ResourceCost } from './economy.ts';

// Each Tech is a BIT POSITION in the owner's u64 techMask (0..63). Stored as u8
// in the `research` table; the mask packs all completed techs into one number so
// combat math reads a single column. New techs append (never renumber — the bit
// index is persisted in player.techMask).
export const Tech = {
  ArmorMail: 0, // foot/missile troops: +1 armor tier (cap Mail) — survive arrows
  ArmorPlate: 1, // melee troops: +hp, harder to cut down
  FletchedArrows: 2, // archers/crossbows: +attack
  SharpenedBlades: 3, // melee attackers: +attack
  Masonry: 4, // structures: +hp / +armor tier (cap Stone)
  Conscription: 5, // ALL combatants: +hp — deeper ranks
} as const;
export type Tech = (typeof Tech)[keyof typeof Tech];

export const ALL_TECHS: Tech[] = [
  Tech.ArmorMail,
  Tech.ArmorPlate,
  Tech.FletchedArrows,
  Tech.SharpenedBlades,
  Tech.Masonry,
  Tech.Conscription,
];

// Additive deltas folded onto a base def. Every field is optional and ADDITIVE
// (armorClass/armorTier bump the tier index, clamped). Absent = no change.
export interface UnitDelta {
  attack?: number; // +flat damage per hit
  maxHp?: number; // +flat hit points
  range?: number; // +reach in world units
  armorTier?: number; // +armor class tiers (clamped to Mail for troops)
}

export interface BuildingDelta {
  maxHp?: number;
  armorTier?: number; // clamped to Stone
}

// Which units a unit-targeting tech touches. A pure predicate over the base def
// (read of UNIT_DEFS), so "applies to every ranged unit" stays data-driven and a
// new ranged unit auto-inherits the bonus with zero tech-table edits.
export type UnitPredicate = (def: UnitDef) => boolean;

const isCombatant = (d: UnitDef): boolean => d.attack > 0;
const isRanged = (d: UnitDef): boolean => !!d.ranged;
const isMelee = (d: UnitDef): boolean =>
  d.attack > 0 && !d.ranged && d.range <= 2;

export interface UpgradeDef {
  label: string;
  icon: string; // emoji shown in the Blacksmith research UI
  cost: ResourceCost;
  researchTime: number; // seconds of progress to complete
  requires?: BuildingKind; // extra building prereq beyond owning a Blacksmith
  appliesTo: UnitPredicate; // which units the unit-delta folds onto
  delta: UnitDelta; // additive bonus applied to every matching unit
  buildingDelta?: BuildingDelta; // optional structural bonus (Masonry)
  appliesToBuildings?: boolean; // when set, buildingDelta folds onto ALL structures
}

// The data. Tuning lives here only — systems read effectiveUnitDef and never see
// a literal bonus. Mail armor is the foot-troop survival pick; plate/sharpened
// split the melee line; fletching is the missile pick; masonry hardens the base;
// conscription is the across-the-board hp tech that always helps.
export const UPGRADE_DEFS: Record<Tech, UpgradeDef> = {
  [Tech.ArmorMail]: {
    label: 'Mail Armor',
    icon: '🥼',
    cost: { wood: 60, gold: 40 },
    researchTime: 30,
    appliesTo: (d) => isCombatant(d) && !d.prefersBuildings, // troops, not siege
    delta: { armorTier: 1 },
  },
  [Tech.ArmorPlate]: {
    label: 'Plate Barding',
    icon: '🛡️',
    cost: { wood: 40, stone: 30, gold: 60 },
    researchTime: 45,
    requires: BuildingKind.Stable,
    appliesTo: (d) => isMelee(d),
    delta: { maxHp: 25 },
  },
  [Tech.FletchedArrows]: {
    label: 'Fletched Arrows',
    icon: '🏹',
    cost: { wood: 50, gold: 30 },
    researchTime: 30,
    appliesTo: (d) => isRanged(d),
    delta: { attack: 3 },
  },
  [Tech.SharpenedBlades]: {
    label: 'Sharpened Blades',
    icon: '⚔️',
    cost: { wood: 50, gold: 30 },
    researchTime: 30,
    appliesTo: (d) => isMelee(d),
    delta: { attack: 3 },
  },
  [Tech.Masonry]: {
    label: 'Masonry',
    icon: '🧱',
    cost: { wood: 40, stone: 80 },
    researchTime: 40,
    appliesTo: () => false, // structures only — no unit effect
    delta: {},
    appliesToBuildings: true,
    buildingDelta: { maxHp: 150, armorTier: 1 },
  },
  [Tech.Conscription]: {
    label: 'Conscription',
    icon: '🪖',
    cost: { food: 60, gold: 50 },
    researchTime: 50,
    requires: BuildingKind.Barracks,
    appliesTo: (d) => isCombatant(d),
    delta: { maxHp: 15 },
  },
};

// ── bitmask helpers ───────────────────────────────────────────────────────────
// The mask is a u64 (bigint). One bit per Tech; bit index == the Tech value.

export function techBit(tech: Tech): bigint {
  return 1n << BigInt(tech);
}

export function hasTech(mask: bigint, tech: Tech): boolean {
  return (mask & techBit(tech)) !== 0n;
}

export function setTech(mask: bigint, tech: Tech): bigint {
  return mask | techBit(tech);
}

// Completed techs in a mask, in canonical (ascending bit) order. Deterministic.
export function techsInMask(mask: bigint): Tech[] {
  return ALL_TECHS.filter((t) => hasTech(mask, t));
}

// ── armor-tier clamping ───────────────────────────────────────────────────────
// Troops cap at Mail (heavy infantry plate is still "mail" class here); structures
// cap at Stone. Tiers never go below the base.

const clampTier = (tier: number, cap: ArmorClass): ArmorClass =>
  Math.max(ArmorClass.Unarmored, Math.min(tier, cap)) as ArmorClass;

// ── effective defs ────────────────────────────────────────────────────────────

// Fold the owner's completed techs (mask) into the base unit def as ADDITIVE
// deltas. Pure and deterministic: same (kind, mask) → byte-identical def, so the
// module (authority) and the client agree, and the bonus auto-applies to current
// AND future units of that kind. Returns a NEW object; never mutates UNIT_DEFS.
export function effectiveUnitDef(kind: number, mask: bigint): UnitDef {
  const base = UNIT_DEFS[kind as 0];
  if (!base || mask === 0n) return base;
  let attack = base.attack;
  let maxHp = base.maxHp;
  let range = base.range;
  let tier = base.armorClass as number;
  let changed = false;
  for (const tech of techsInMask(mask)) {
    const up = UPGRADE_DEFS[tech];
    if (!up.appliesTo(base)) continue;
    const d = up.delta;
    if (d.attack) attack += d.attack;
    if (d.maxHp) maxHp += d.maxHp;
    if (d.range) range += d.range;
    if (d.armorTier) tier += d.armorTier;
    changed = true;
  }
  if (!changed) return base;
  return {
    ...base,
    attack,
    maxHp,
    range,
    armorClass: clampTier(tier, ArmorClass.Mail),
  };
}

// Fold structural techs (Masonry) into a base building def. Same purity contract.
export function effectiveBuildingDef(kind: number, mask: bigint): BuildingDef {
  const base = BUILDING_DEFS[kind as 0];
  if (!base || mask === 0n) return base;
  let maxHp = base.maxHp;
  let tier = base.armorClass as number;
  let changed = false;
  for (const tech of techsInMask(mask)) {
    const up = UPGRADE_DEFS[tech];
    if (!up.appliesToBuildings || !up.buildingDelta) continue;
    const d = up.buildingDelta;
    if (d.maxHp) maxHp += d.maxHp;
    if (d.armorTier) tier += d.armorTier;
    changed = true;
  }
  if (!changed) return base;
  return { ...base, maxHp, armorClass: clampTier(tier, ArmorClass.Stone) };
}

File: saladin-bevy/crates/sim/src/research.rs
use crate::buildings_defs::{BuildingDef, building_def};
use crate::economy::{ResourceCost, Stockpile};
use crate::enums::{ArmorClass, BuildingKind};
use crate::math::Fx;
use crate::tech::has_prereq;
use crate::units::{UnitDef, unit_def};
use crate::enums::UnitKind;
use serde::{Deserialize, Serialize};
use std::collections::HashSet;

/// Blacksmith research. Each `Tech` is a bit position in the owner's u64
/// `tech_mask`. Bonuses are NEVER baked onto rows; they are DERIVED on read via
/// `effective_unit_def` / `effective_building_def`, so an upgrade applies to
/// every current and future unit of a kind automatically.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[repr(u8)]
pub enum Tech {
    ArmorMail = 0,
    ArmorPlate = 1,
    FletchedArrows = 2,
    SharpenedBlades = 3,
    Masonry = 4,
    Conscription = 5,
}

pub const ALL_TECHS: [Tech; 6] = [
    Tech::ArmorMail,
    Tech::ArmorPlate,
    Tech::FletchedArrows,
    Tech::SharpenedBlades,
    Tech::Masonry,
    Tech::Conscription,
];

impl Tech {
    pub fn from_u8(v: u8) -> Option<Tech> {
        ALL_TECHS.iter().copied().find(|t| *t as u8 == v)
    }
}

/// Additive deltas folded onto a base def (0 == no change).
#[derive(Clone, Copy, Debug)]
pub struct UnitDelta {
    pub attack: i32,
    pub max_hp: i32,
    pub range: Fx,
    pub armor_tier: i32,
}

const NO_DELTA: UnitDelta = UnitDelta { attack: 0, max_hp: 0, range: crate::fx!("0"), armor_tier: 0 };

#[derive(Clone, Copy, Debug)]
pub struct BuildingDelta {
    pub max_hp: i32,
    pub armor_tier: i32,
}

#[derive(Clone, Copy, Debug)]
pub struct UpgradeDef {
    pub label: &'static str,
    pub icon: &'static str,
    pub cost: ResourceCost,
    pub research_time: Fx,
    pub requires: Option<BuildingKind>,
    pub applies_to: fn(&UnitDef) -> bool,
    pub delta: UnitDelta,
    pub building_delta: Option<BuildingDelta>,
    pub applies_to_buildings: bool,
}

fn is_combatant(d: &UnitDef) -> bool {
    d.attack > 0
}
fn is_ranged(d: &UnitDef) -> bool {
    d.ranged
}
fn is_melee(d: &UnitDef) -> bool {
    d.attack > 0 && !d.ranged && d.range <= crate::fx!("2")
}
fn troops_not_siege(d: &UnitDef) -> bool {
    is_combatant(d) && !d.prefers_buildings
}
fn never(_: &UnitDef) -> bool {
    false
}

const UPGRADE_DEFS: [UpgradeDef; 6] = [
    // ArmorMail
    UpgradeDef {
        label: "Mail Armor",
        icon: "🥼",
        cost: ResourceCost::new(60, 0, 0, 40),
        research_time: crate::fx!("30"),
        requires: None,
        applies_to: troops_not_siege,
        delta: UnitDelta { armor_tier: 1, ..NO_DELTA },
        building_delta: None,
        applies_to_buildings: false,
    },
    // ArmorPlate
    UpgradeDef {
        label: "Plate Barding",
        icon: "🛡️",
        cost: ResourceCost::new(40, 30, 0, 60),
        research_time: crate::fx!("45"),
        requires: Some(BuildingKind::Stable),
        applies_to: is_melee,
        delta: UnitDelta { max_hp: 25, ..NO_DELTA },
        building_delta: None,
        applies_to_buildings: false,
    },
    // FletchedArrows
    UpgradeDef {
        label: "Fletched Arrows",
        icon: "🏹",
        cost: ResourceCost::new(50, 0, 0, 30),
        research_time: crate::fx!("30"),
        requires: None,
        applies_to: is_ranged,
        delta: UnitDelta { attack: 3, ..NO_DELTA },
        building_delta: None,
        applies_to_buildings: false,
    },
    // SharpenedBlades
    UpgradeDef {
        label: "Sharpened Blades",
        icon: "⚔️",
        cost: ResourceCost::new(50, 0, 0, 30),
        research_time: crate::fx!("30"),
        requires: None,
        applies_to: is_melee,
        delta: UnitDelta { attack: 3, ..NO_DELTA },
        building_delta: None,
        applies_to_buildings: false,
    },
    // Masonry
    UpgradeDef {
        label: "Masonry",
        icon: "🧱",
        cost: ResourceCost::new(40, 80, 0, 0),
        research_time: crate::fx!("40"),
        requires: None,
        applies_to: never,
        delta: NO_DELTA,
        building_delta: Some(BuildingDelta { max_hp: 150, armor_tier: 1 }),
        applies_to_buildings: true,
    },
    // Conscription
    UpgradeDef {
        label: "Conscription",
        icon: "🪖",
        cost: ResourceCost::new(0, 0, 60, 50),
        research_time: crate::fx!("50"),
        requires: Some(BuildingKind::Barracks),
        applies_to: is_combatant,
        delta: UnitDelta { max_hp: 15, ..NO_DELTA },
        building_delta: None,
        applies_to_buildings: false,
    },
];

pub fn upgrade_def(tech: Tech) -> &'static UpgradeDef {
    &UPGRADE_DEFS[tech as usize]
}

// ── bitmask ─────────────────────────────────────────────────────────────────

pub fn tech_bit(tech: Tech) -> u64 {
    1u64 << (tech as u8)
}
pub fn has_tech(mask: u64, tech: Tech) -> bool {
    mask & tech_bit(tech) != 0
}
pub fn set_tech(mask: u64, tech: Tech) -> u64 {
    mask | tech_bit(tech)
}
pub fn techs_in_mask(mask: u64) -> Vec<Tech> {
    ALL_TECHS.iter().copied().filter(|t| has_tech(mask, *t)).collect()
}

fn clamp_tier(tier: i32, cap: ArmorClass) -> ArmorClass {
    let t = tier.clamp(0, cap as i32) as u8;
    ArmorClass::from_u8(t).unwrap_or(ArmorClass::Unarmored)
}

/// Fold the owner's completed techs into the base unit def as additive deltas.
/// Pure: same `(kind, mask)` → identical def. Predicate reads the BASE def.
pub fn effective_unit_def(kind: UnitKind, mask: u64) -> UnitDef {
    let base = *unit_def(kind);
    if mask == 0 {
        return base;
    }
    let mut out = base;
    let mut tier = base.armor_class as i32;
    let mut changed = false;
    for tech in techs_in_mask(mask) {
        let up = upgrade_def(tech);
        if !(up.applies_to)(&base) {
            continue;
        }
        out.attack += up.delta.attack;
        out.max_hp += up.delta.max_hp;
        out.range += up.delta.range;
        tier += up.delta.armor_tier;
        changed = true;
    }
    if !changed {
        return base;
    }
    out.armor_class = clamp_tier(tier, ArmorClass::Mail);
    out
}

/// Fold structural techs (Masonry) into a base building def.
pub fn effective_building_def(kind: BuildingKind, mask: u64) -> BuildingDef {
    let base = *building_def(kind);
    if mask == 0 {
        return base;
    }
    let mut out = base;
    let mut tier = base.armor_class as i32;
    let mut changed = false;
    for tech in techs_in_mask(mask) {
        let up = upgrade_def(tech);
        let Some(d) = up.building_delta else { continue };
        if !up.applies_to_buildings {
            continue;
        }
        out.max_hp += d.max_hp;
        tier += d.armor_tier;
        changed = true;
    }
    if !changed {
        return base;
    }
    out.armor_class = clamp_tier(tier, ArmorClass::Stone);
    out
}

// ── research panel (UI-facing, pure) ─────────────────────────────────────────

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum ResearchStatus {
    Done,
    InProgress,
    Locked,
    Unaffordable,
    Available,
}

/// Minimal shape of a research-table row the panel needs.
#[derive(Clone, Copy, Debug)]
pub struct ResearchProgressRow {
    pub tech: u8,
    pub progress: Fx,
    pub done: bool,
}

#[derive(Clone, Debug)]
pub struct ResearchRowState {
    pub tech: Tech,
    pub label: &'static str,
    pub icon: &'static str,
    pub cost: ResourceCost,
    pub status: ResearchStatus,
    pub progress: Fx,
    pub lock_note: Option<String>,
}

/// One descriptor per tech for the Blacksmith research panel. Status precedence:
/// done > in_progress > locked > unaffordable > available.
pub fn research_panel_state(
    mask: u64,
    rows: &[ResearchProgressRow],
    stock: &Stockpile,
    owned_buildings: &HashSet<BuildingKind>,
) -> Vec<ResearchRowState> {
    ALL_TECHS
        .iter()
        .copied()
        .map(|tech| {
            let up = upgrade_def(tech);
            let row = rows.iter().find(|r| r.tech == tech as u8);
            let mk = |status, progress, lock_note| ResearchRowState {
                tech,
                label: up.label,
                icon: up.icon,
                cost: up.cost,
                status,
                progress,
                lock_note,
            };

            if has_tech(mask, tech) {
                return mk(ResearchStatus::Done, Fx::ONE, None);
            }
            if let Some(r) = row {
                if !r.done {
                    let p = r.progress.clamp(Fx::ZERO, Fx::ONE);
                    return mk(ResearchStatus::InProgress, p, None);
                }
            }
            if let Some(req) = up.requires {
                if !has_prereq(owned_buildings, Some(req)) {
                    let note = format!("Requires {}", building_def(req).label);
                    return mk(ResearchStatus::Locked, Fx::ZERO, Some(note));
                }
            }
            if !stock.can_afford(&up.cost) {
                return mk(ResearchStatus::Unaffordable, Fx::ZERO, None);
            }
            mk(ResearchStatus::Available, Fx::ZERO, None)
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mask_bits() {
        let m = set_tech(0, Tech::FletchedArrows);
        assert!(has_tech(m, Tech::FletchedArrows));
        assert!(!has_tech(m, Tech::ArmorMail));
        assert_eq!(techs_in_mask(set_tech(m, Tech::ArmorMail)), vec![Tech::ArmorMail, Tech::FletchedArrows]);
    }

    #[test]
    fn fletched_arrows_boosts_only_ranged() {
        let m = set_tech(0, Tech::FletchedArrows);
        assert_eq!(effective_unit_def(UnitKind::Archer, m).attack, 12); // 9 + 3
        assert_eq!(effective_unit_def(UnitKind::Spearman, m).attack, 12); // unchanged (melee)
    }

    #[test]
    fn mail_armor_bumps_tier_clamped() {
        let m = set_tech(0, Tech::ArmorMail);
        // Archer Leather(1) -> Mail(2)
        assert_eq!(effective_unit_def(UnitKind::Archer, m).armor_class, ArmorClass::Mail);
        // Knight already Mail(2) -> clamp keeps Mail
        assert_eq!(effective_unit_def(UnitKind::Knight, m).armor_class, ArmorClass::Mail);
    }

    #[test]
    fn masonry_hardens_buildings_only() {
        let m = set_tech(0, Tech::Masonry);
        let keep = effective_building_def(BuildingKind::Keep, m);
        assert_eq!(keep.max_hp, 1650); // 1500 + 150
        // unit unaffected
        assert_eq!(effective_unit_def(UnitKind::Spearman, m).max_hp, 70);
    }

    #[test]
    fn panel_precedence() {
        let owned: HashSet<BuildingKind> = HashSet::new();
        let rich = Stockpile { wood: 999, stone: 999, food: 999, gold: 999 };
        let rows: Vec<ResearchProgressRow> = vec![];
        let panel = research_panel_state(0, &rows, &rich, &owned);
        // ArmorPlate requires Stable (not owned) -> Locked
        let plate = panel.iter().find(|r| r.tech == Tech::ArmorPlate).unwrap();
        assert_eq!(plate.status, ResearchStatus::Locked);
        // ArmorMail has no prereq, affordable -> Available
        let mail = panel.iter().find(|r| r.tech == Tech::ArmorMail).unwrap();
        assert_eq!(mail.status, ResearchStatus::Available);
    }
}
